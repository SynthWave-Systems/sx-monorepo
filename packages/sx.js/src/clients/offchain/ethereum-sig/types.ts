export const domain = {
  name: 'snapshot',
  version: '0.1.4'
};

export const voteTypes = {
  Vote: [
    { name: 'from', type: 'address' },
    { name: 'space', type: 'string' },
    { name: 'timestamp', type: 'uint64' },
    { name: 'proposal', type: 'string' },
    { name: 'choice', type: 'uint32' },
    { name: 'reason', type: 'string' },
    { name: 'app', type: 'string' },
    { name: 'metadata', type: 'string' }
  ]
};
