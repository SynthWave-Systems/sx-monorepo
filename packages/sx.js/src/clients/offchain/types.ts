import type { TypedDataDomain, TypedDataField } from '@ethersproject/abstract-signer';

enum Choice {
  Against = 2,
  For = 1,
  Abstain = 3
}

export type SignatureData = {
  address: string;
  signature: string;
  domain: TypedDataDomain;
  types: Record<string, TypedDataField[]>;
  message: Record<string, any>;
};

export type Envelope<T extends Vote> = {
  signatureData?: SignatureData;
  data: T;
};

export type StrategyConfig = {
  index: number;
  address: string;
  metadata?: Record<string, any>;
};

export type EIP712VoteMessage = {
  space: string;
  proposal: string;
  choice: number | number[] | string | { [key: string]: number };
  reason: string;
  app: string;
  metadata: string;
};

export type EIP712Message = EIP712VoteMessage & { timestamp: number; from: string };

export type Vote = {
  space: string;
  authenticator: string;
  strategies: StrategyConfig[];
  proposal: string;
  choice: Choice;
  metadataUri: string;
};
