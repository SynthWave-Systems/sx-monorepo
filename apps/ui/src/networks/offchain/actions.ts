import { StrategyParsedMetadata } from '@/types';
import {
  ReadOnlyNetworkActions,
  NetworkConstants,
  NetworkHelpers,
  SnapshotInfo,
  VotingPower
} from '../types';

const SCORE_URL = 'https://score.snapshot.org';

export function createActions(
  constants: NetworkConstants,
  helpers: NetworkHelpers,
  chainId: number
): ReadOnlyNetworkActions {
  return {
    getVotingPower: async (
      strategiesAddresses: string[],
      strategiesParams: any[],
      strategiesMetadata: StrategyParsedMetadata[],
      voterAddress: string,
      snapshotInfo: SnapshotInfo
    ): Promise<VotingPower[]> => {
      const result = await fetch(SCORE_URL, {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          method: 'get_vp',
          params: {
            address: voterAddress,
            space: '',
            strategies: strategiesParams,
            network: snapshotInfo.chainId ?? chainId,
            snapshot: snapshotInfo.at ?? 'latest'
          }
        })
      });
      const body = await result.json();

      return body.result.vp_by_strategy.map((vp: number, index: number) => {
        const strategy = strategiesParams[index];
        const decimals = parseInt(strategy.params.decimals || 0);

        return {
          address: strategy.name,
          value: BigInt(vp * 10 ** decimals),
          decimals,
          symbol: strategy.params.symbol,
          token: strategy.params.address,
          chainId: strategy.network ? parseInt(strategy.network) : undefined
        } as VotingPower;
      });
    }
  };
}
