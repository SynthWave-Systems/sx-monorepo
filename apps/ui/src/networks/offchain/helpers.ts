import { Choice } from '@/types';

export function getSdkChoice(choice: Choice): number {
  if (choice === 'for') return 1;
  if (choice === 'against') return 2;
  return 3;
}
