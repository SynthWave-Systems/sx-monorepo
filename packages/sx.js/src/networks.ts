export * from './evmNetworks';
export * from './starknetNetworks';
export * from './offchainNetworks';
