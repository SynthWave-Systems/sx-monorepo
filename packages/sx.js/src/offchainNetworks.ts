function createStandardConfig(eip712ChainId: number) {
  return {
    eip712ChainId
  };
}

export const offchainMainnet = createStandardConfig(1);
export const offchainGoerli = createStandardConfig(5);
