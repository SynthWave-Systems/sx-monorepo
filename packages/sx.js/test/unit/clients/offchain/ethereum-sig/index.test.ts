import { Wallet } from '@ethersproject/wallet';
import { EthereumSig } from '../../../../../src/clients/offchain/ethereum-sig';
import { offchainGoerli } from '../../../../../src/offchainNetworks';
import { voteTypes, domain } from '../../../../../src/clients/offchain/ethereum-sig/types';

describe('EthereumSig', () => {
  // Test address: 0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90
  const TEST_PK = 'ef4bcf36b5d026b703b86a311031fe2291b979620f01443f795fa213f9105e35';
  const signer = new Wallet(TEST_PK);
  const client = new EthereumSig({ networkConfig: offchainGoerli });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should create vote envelope', async () => {
    const payload = {
      space: 'wan-test.eth',
      authenticator: '',
      strategies: [],
      proposal: '0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801',
      choice: 1,
      metadataUri: ''
    };
    const address = await signer.getAddress();

    const envelope = await client.vote({
      signer,
      data: payload
    });

    expect(envelope).toEqual({
      data: payload,
      signatureData: {
        signature: expect.any(String),
        address,
        types: voteTypes,
        domain,
        message: {
          app: '',
          choice: payload.choice,
          from: address,
          metadata: '',
          proposal: payload.proposal,
          reason: '',
          space: payload.space,
          timestamp: expect.any(Number)
        }
      }
    });
  });
});
