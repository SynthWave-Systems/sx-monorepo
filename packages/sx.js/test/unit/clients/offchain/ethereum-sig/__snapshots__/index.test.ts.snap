// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumSig should create vote envelope 1`] = `
{
  "data": {
    "authenticator": "",
    "choice": 1,
    "metadataUri": "",
    "proposal": "0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801",
    "space": "wan-test.eth",
    "strategies": [],
  },
  "signatureData": {
    "address": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
    "domain": {
      "name": "snapshot",
      "version": "0.1.4",
    },
    "message": {
      "app": "",
      "choice": 1,
      "from": "0xf1f09AdC06aAB740AA16004D62Dbd89484d3Be90",
      "metadata": "",
      "proposal": "0xcc47146e5e0ac781e8976405a8da4468f2a0c4cdf0c7659353d728fafe46f801",
      "reason": "",
      "space": "wan-test.eth",
      "timestamp": 1708024006,
    },
    "signature": "0x6b7d82de78c37d67587fa7554222e6e9779927b1ea87e6926e1a76dbf94a87c83e55d70333ff860fc7c393e634e39b20eccf1d21c2a7a111f8bafcced1bfdde01c",
    "types": {
      "Vote": [
        {
          "name": "from",
          "type": "address",
        },
        {
          "name": "space",
          "type": "string",
        },
        {
          "name": "timestamp",
          "type": "uint64",
        },
        {
          "name": "proposal",
          "type": "string",
        },
        {
          "name": "choice",
          "type": "uint32",
        },
        {
          "name": "reason",
          "type": "string",
        },
        {
          "name": "app",
          "type": "string",
        },
        {
          "name": "metadata",
          "type": "string",
        },
      ],
    },
  },
}
`;
